### PR TITLE
Package updates

### PR DESCRIPTION
**Pull request by nabobalis**
1. Updated tox by copying astropy
2. Added basic pre-commit for people who want to use one.
3. Removed version.py by again coping astropy.

~Probs more to do.~

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+repos:
+-   repo: https://github.com/pre-commit/mirrors-autopep8
```
Comment (Cadair) [inline .pre-commit-config.yaml]: Does this end up applying to the whole file when you commit it? Does that make it likely to generate large style diffs?

Comment (nabobalis) [reply .pre-commit-config.yaml]: Yes. I would suggest we don't allow anyone to use it until we lint the entire package. But I thought to add it now.

File: setup.cfg [hunk 2]
```diff
@@ -102,23 +112,35 @@ markers =
...
 [flake8]
-max-line-length = 100
+max-line-length = 110
+exclude =
+    .git,
+    __pycache__,
+    docs/conf.py,
+    build,
+    sunpy/data/sample.py,
+    sunpy/extern/distro.py,
```
Comment (Cadair) [inline setup.cfg]: can we ignore a whole directory?

Comment (nabobalis) [reply setup.cfg]: That makes more sense, I will check and add it.

File: setup.py [hunk 3]
```diff
@@ -40,12 +45,17 @@
...
-from setuptools.config import read_configuration
-
-from extension_helpers import get_extensions
+VERSION_TEMPLATE = """
+# Note that we need to fall back to the hard-coded version if either
+# setuptools_scm can't be imported or setuptools_scm can't determine the
+# version, so we catch the generic 'Exception'.
+try:
+    from setuptools_scm import get_version
+    version = get_version(root='..', relative_to=__file__)
```
Comment (Cadair) [inline setup.py]: An interesting suggestion from the astropy discussion here is to change this to:
```suggestion
    __version__ = get_version(root='..', relative_to=__file__)
```

and then update `__init__.py` so it does `from version import __version__`.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (Cadair) [reply .pre-commit-config.yaml]: hmmm, I wonder if this is useful before then anyway.

Comment (nabobalis) [reply .pre-commit-config.yaml]: If you dont have pre commit installed it wont do anything or tell you its here.

Comment (Cadair) [reply .pre-commit-config.yaml]: I was more meaning people could install it and use it before we apply it to all the files in the repo, for isort and the more basic checks.

File: sunpy/__init__.py [hunk 4]
```diff
@@ -31,8 +32,6 @@ class UnsupportedPythonError(Exception):
     import builtins
     builtins._SUNPY_SETUP_ = False
```
Comment (Cadair) [inline sunpy/__init__.py]: I wonder if we can remove this?

File: tox.ini [hunk 5]
```diff
@@ -1,76 +1,147 @@
...
+# E722 - do not use bare except
+# E901 - SyntaxError or IndentationError
+# E902 - IOError
+# E999: SyntaxError -- failed to compile a file into an Abstract Syntax Tree
+# F822: undefined name in __all__
+# F823: local variable name referenced before assignment
+skip_install = true
+description = check code style, e.g. with flake8
+deps = flake8
+commands = flake8 sunpy --count --select=E101,W191,W291,W292,W293,W391,E111,E112,E113,E30,E502,E722,E901,E902,E999,F822,F823
```
Comment (Cadair) [inline tox.ini]: Should we limit the pre-commit flake 8 to this list of checks?

File: setup.cfg [hunk 6]
```diff
@@ -67,6 +67,16 @@ docs =
   sphinx-gallery
   sunpy-sphinx-theme
   towncrier
+  reproject
+  astroquery
+  jplephem
+format =
```
Comment (Cadair) [inline setup.cfg]: Are we using this for anything?

File: .pre-commit-config.yaml (hunk 1 above)
Comment (nabobalis) [reply .pre-commit-config.yaml]: Sure but if we never tell people it exists they wont hopefully?

File: sunpy/__init__.py (hunk 4 above)
Comment (nabobalis) [reply sunpy/__init__.py]: Let me check.

File: tox.ini (hunk 5 above)
Comment (nabobalis) [reply tox.ini]: I mean I can? I thought the idea was that we check a subset on the CI but make IDE users sort all problems?

File: setup.cfg (hunk 6 above)
Comment (nabobalis) [reply setup.cfg]: No I just thought that pydoc style would use this in the future to check th entire code base using all these packages.

File: sunpy/__init__.py (hunk 4 above)
Comment (nabobalis) [reply sunpy/__init__.py]: I've removed it, it seems to work locally.

File: setup.cfg (hunk 6 above)
Comment (Cadair) [reply setup.cfg]: Is it possible to make pre-commit do all this?

Comment (nabobalis) [reply setup.cfg]: In the future I don't see why not.

**Comment by nabobalis** [comment]
**SQUASH MERGE THIS**

File: tox.ini (hunk 5 above)
Comment (Cadair) [reply tox.ini]: I like the idea of having the pre-commit be a way to make the style CI pass.

Comment (nabobalis) [reply tox.ini]: Already done.

File: setup.cfg (hunk 6 above)
Comment (Cadair) [reply setup.cfg]: I meant we can make pre-commit install these things, so do we really need a new extra for them?

File: tox.ini [hunk 7]
```diff
@@ -1,76 +1,124 @@
 [tox]
-envlist = py{36,37,38}-{offline},online,build_docs,figure
-isolated_build = True
+envlist =
+    py{36,37,38}-test{,-oldestdeps,-devdeps,-online}
```
Comment (Cadair) [inline tox.ini]: are we doing anything with `-test`? Is is clearer to have `py38-test` than just `py38`?

File: tox.ini [hunk 8]
```diff
@@ -1,76 +1,124 @@
...
     pytest-xdist
-    astropydev: cython
-    astropydev,figure_astropydev: git+https://github.com/astropy/astropy
-    astropy31: astropy<3.2
+
+    # The devdeps factor is intended to be used to install the latest developer version.
+    # of key dependencies.
+    devdeps: cython
+    devdeps,figure_astropydev: git+https://github.com/astropy/astropy
+    devdeps: astropy<3.2
```
Comment (Cadair) [inline tox.ini]: wait how are we installing git astropy and astropy <3.2?!

File: tox.ini [hunk 9]
```diff
@@ -1,76 +1,124 @@
...
+
+    # All our tests run in parallel which this plugin allows us to.
     pytest-xdist
-    astropydev: cython
-    astropydev,figure_astropydev: git+https://github.com/astropy/astropy
-    astropy31: astropy<3.2
+
+    # The devdeps factor is intended to be used to install the latest developer version.
+    # of key dependencies.
+    devdeps: cython
```
Comment (Cadair) [inline tox.ini]: Once astropy/astropy#9726 is merged we probably don't need this any more.

File: tox.ini (hunk 8 above)
Comment (Cadair) [reply tox.ini]: Also, we should add some more oldest deps to this, like mpl, numpy whatever else we pin in setup.cfg.

File: setup.cfg (hunk 6 above)
Comment (nabobalis) [reply setup.cfg]: Oh I thought I removed them.

File: tox.ini (hunk 7 above)
Comment (nabobalis) [reply tox.ini]: Nope, i can remove it.

File: tox.ini (hunk 8 above)
Comment (nabobalis) [reply tox.ini]: Woops.

Comment (nabobalis) [reply tox.ini]: Woops.

File: tox.ini (hunk 7 above)
Comment (nabobalis) [reply tox.ini]: Removed.

File: setup.cfg (hunk 6 above)
Comment (nabobalis) [reply setup.cfg]: Removed.

File: sunpy/__init__.py [hunk 10]
```diff
@@ -9,6 +9,12 @@
 """
 # Enforce Python version check during package import.
 # This is the same check as the one at the top of setup.py
+from sunpy.util.logger import _init_log
```
Comment (Cadair) [inline sunpy/__init__.py]: isort?! :stuck_out_tongue:

File: tox.ini [hunk 11]
```diff
@@ -1,76 +1,128 @@
 [tox]
-envlist = py{36,37,38}-{offline},online,build_docs,figure
-isolated_build = True
+envlist =
+    py{36,37,38}{,-oldestdeps,-devdeps,-online}
+    build_docs
+    figure
+    figure_astropydev
+    conda
```
Comment (Cadair) [inline tox.ini]:
```suggestion
    py36-conda
```
At least this is what the env is further down?

File: sunpy/__init__.py (hunk 10 above)
Comment (nabobalis) [reply sunpy/__init__.py]: Fixed,

File: tox.ini (hunk 11 above)
Comment (nabobalis) [reply tox.ini]: Fixed

**Comment by nabobalis** [comment]
Should check that before we merge this.

**Comment by Cadair** [comment]
It looks like it could well happen, but only with an old version of asdf. Not sure what's happening there.

**Comment by nabobalis** [comment]
Can we delete asdf and just use a real file format like GeoTIFF?

**Pull request merged**